### DISCUSSION TITLE
Data Consistency of `RequestBeginBlock`

### DISCUSSION THREAD

**Discussion by faustbrian** [tendermint/tendermint#9526] [Q&A] [answered]
So my question is about the `RequestBeginBlock` proto message. Is it safe to assume that this message is the same across all nodes and can be made part of the `appHash` that is returned in `Commit` or does this value run the risk of being different across nodes and still reaching consensus?

My use-case for making it part of the app hash would be that the application does a fair few things that rely on data from `RequestBeginBlock` so making sure that everyone has the same `RequestBeginBlock` payload for the last committed block would add some extra proof of data consistency across nodes.

**Comment by sergio-mena** [comment, marked as answer]
All fields in `RequestBeginBlock` are part of the newly produced (or "decided") block, so these values MUST be the same at all nodes (by Agreement property of Tendermint consensus algorithm).

HOWEVER, be careful if you are planning to use a protobuf-marshalled version of a received `RequestBeginBlock` as input for your `appHash`. I did a quick lookup and found [this](https://gist.github.com/kchristidis/39c8b310fd9da43d515c4394c3cd9510).

ABCI specification does not require any guarantees or make any assumptions on the determinism of protobuf serialization. Agreement MUST hold for the _values_ of the decided block's fields.

**Reply by faustbrian** [reply]
Was primarily concerned about the values since determinism in serialization is fixable but a lack of determinism in the data would be more difficult to deal with 😅
